istioctl/cmd/workload.go

```diff
@@ -0,0 +1,402 @@
+func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, clusterID, revision, dir string) (*meshconfig.MeshConfig, error) {
+	istioCM := "istio"
+	// Case with multiple control planes
+	if revision != "" {
+		istioCM = fmt.Sprintf("%s-%s", istioCM, revision)
+	}
+	istio, err := kubeClient.CoreV1().ConfigMaps(istioNamespace).Get(context.Background(), istioCM, metav1.GetOptions{})
+	// errors if the requested configmap does not exist in the given namespace
+	if err != nil {
+		return nil, fmt.Errorf("configmap %s was not found in namespace %s: %v", istioCM, istioNamespace, err)
+	}
+	meshConfig, err := mesh.ApplyMeshConfigDefaults(istio.Data[configMapKey])
+	if err != nil {
+		return nil, err
+	}
+
+	labels := map[string]string{}
+	for k, v := range wg.Spec.Metadata.Labels {
+		labels[k] = v
+	}
+	// case where a user provided custom workload group has labels in the workload entry template field
+	for k, v := range wg.Spec.Template.Labels {
+		labels[k] = v
+	}
+	we := wg.Spec.Template
+	md := meshConfig.DefaultConfig.ProxyMetadata
+	md["CANONICAL_SERVICE"], md["CANONICAL_REVISION"] = inject.ExtractCanonicalServiceLabels(labels, wg.Name)
+	md["DNS_AGENT"] = ""
+	md["POD_NAMESPACE"] = wg.Namespace
+	md["SERVICE_ACCOUNT"] = we.ServiceAccount
+	md["TRUST_DOMAIN"] = meshConfig.TrustDomain
+
+	md["ISTIO_META_CLUSTER_ID"] = clusterID
+	md["ISTIO_META_MESH_ID"] = meshConfig.DefaultConfig.MeshId
+	md["ISTIO_META_NETWORK"] = we.Network
+	if portsJSON, err := json.Marshal(we.Ports); err == nil {
+		md["ISTIO_META_POD_PORTS"] = string(portsJSON)
+	}
+	md["ISTIO_META_WORKLOAD_NAME"] = wg.Name
+	labels["service.istio.io/canonical-name"] = md["CANONICAL_SERVICE"]
+	labels["service.istio.io/canonical-version"] = md["CANONICAL_REVISION"]
+	if labelsJSON, err := json.Marshal(labels); err == nil {
+		md["ISTIO_METAJSON_LABELS"] = string(labelsJSON)
+	}
+
+	meshYAML, err := gogoprotomarshal.ToYAML(meshConfig)
+	if err != nil {
+		return nil, err
+	}
+	return meshConfig, ioutil.WriteFile(filepath.Join(dir, "mesh"), []byte(meshYAML), filePerms)
+}
+
+// Retrieves the external IP of the ingress-gateway for the hosts file additions
+func createHosts(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, meshConfig *meshconfig.MeshConfig, dir string) error {
+	ingress, err := kubeClient.CoreV1().Services(istioNamespace).Get(context.Background(), multicluster.IstioIngressGatewayServiceName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s was not found in namespace %s: %v", multicluster.IstioIngressGatewayServiceName, istioNamespace, err)
+	}
+	serviceIP := ingress.Status.LoadBalancer.Ingress[0].IP
+
+	var hosts []string
+	hosts = append(hosts, fmt.Sprintf("%s istiod.%s.svc\n", serviceIP, istioNamespace),
+		fmt.Sprintf("1.1.1.1 pod.%s.svc.%s\n", wg.Namespace, meshConfig.DefaultConfig.MeshId),
+		fmt.Sprintf("1.1.1.1 vm.%s.svc.%s\n", wg.Namespace, meshConfig.DefaultConfig.MeshId))
```

Comment:
curious: where did `1.1.1.1` come from? is that the current recommendation?

Reply:
Can you remove this? It was a hack use in the integration tests because we did not have DNS for VMs. Real DNS will require either:
* In proxy DNS
* DNS forwarding to kube-dns (or other dns that is kube aware)
* (worst option) manually adding VIPs per k8s service we talk to

I don't think there is value in hardcoding a fake VIP for hardcoded services `pod` and `vm`